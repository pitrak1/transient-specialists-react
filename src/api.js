import axios from 'axios'
import utils from './utils'

const attachInterceptors = () => {
  axios.interceptors.response.use(
    function(response) {
      return response.data.statusCode === 500
        ? Promise.reject(response)
        : response
    },
    function(error) {
      return Promise.reject(error)
    },
  )
}

const translateError = error => {
  if (error.includes('duplicate key value violates unique constraint')) {
    if (error.includes('idx_type_name')) {
      return 'Name must be unique (case insensitive)'
    }

    if (error.includes('idx_oem_name')) {
      return 'Name must be unique (case insensitive)'
    }

    if (error.includes('idx_model_name')) {
      return 'Name must be unique (case insensitive)'
    }

    if (error.includes('idx_equipment_serial_number')) {
      return 'Serial Number must be unique (case insensitive)'
    }
  }

  if (error.includes('violates foreign key constraint')) {
    if (error.includes('equipments_model_id_fkey')) {
      return 'Model cannot be deleted because it has related equipment'
    }

    if (error.includes('equipments_type_id_fkey')) {
      return 'Type cannot be deleted because it has related equipment'
    }

    if (error.includes('models_oem_id_fkey')) {
      return 'OEM cannot be deleted because it has related models'
    }
  }

  return error
}

const getIndex = (resource, options, success, failure) => {
  attachInterceptors()
  const { ascending, page, perPage, searchValue, sortBy } = options
  return axios
    .get(
      `${
        process.env.LAMBDA_ENDPOINT
      }${resource}?ascending=${ascending}&page=${page}&perPage=${perPage}&searchValue=${searchValue}&sortBy=${utils.pascalToSnake(
        sortBy,
      )}`,
    )
    .then(result => {
      success(utils.convertObject(result.data.body))
    })
    .catch(error => {
      failure(error.data.body)
    })
}

const getShow = (resource, id, options, success, failure) => {
  attachInterceptors()
  const { ascending, page, perPage, sortBy } = options
  return axios
    .get(
      `${
        process.env.LAMBDA_ENDPOINT
      }${resource}?id=${id}&ascending=${ascending}&page=${page}&perPage=${perPage}&sortBy=${utils.pascalToSnake(
        sortBy,
      )}`,
    )
    .then(result => {
      success(utils.convertObject(result.data.body))
    })
    .catch(error => {
      failure(error.data.body)
    })
}

const getNew = (resource, success, failure) => {
  attachInterceptors()
  return axios
    .get(`${process.env.LAMBDA_ENDPOINT}${resource}?new=true`)
    .then(result => {
      success(utils.convertObject(result.data.body))
    })
    .catch(error => {
      failure(error.data.body)
    })
}

export const postCreate = (resource, data, success, failure) => {
  attachInterceptors()
  return axios
    .post(`${process.env.LAMBDA_ENDPOINT}${resource}`, data)
    .then(response => {
      success(response.data.body)
    })
    .catch(error => {
      failure(translateError(error.data.body))
    })
}

export const patchUpdate = (resource, data, success, failure) => {
  attachInterceptors()
  return axios
    .patch(`${process.env.LAMBDA_ENDPOINT}${resource}`, data)
    .then(response => {
      success(response.data.body)
    })
    .catch(error => {
      failure(translateError(error.data.body))
    })
}

export const deleteDestroy = (resource, id, success, failure) => {
  attachInterceptors()
  return axios
    .delete(`${process.env.LAMBDA_ENDPOINT}${resource}?id=${id}`)
    .then(response => {
      success(response.data.body)
    })
    .catch(error => {
      failure(translateError(error.data.body))
    })
}

export default {
  getIndex,
  getShow,
  getNew,
  postCreate,
  patchUpdate,
  deleteDestroy,
}
