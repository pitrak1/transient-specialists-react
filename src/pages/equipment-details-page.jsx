import React from 'react'
import {
  Button,
  CircularProgress,
  Grid,
  Toolbar,
  Typography,
} from '@material-ui/core'
import ErrorAlert from '../components/error-alert.jsx'
import Spinner from '../components/spinner.jsx'
import Title from '../components/title.jsx'
import FullTable from '../components/table/full-table.jsx'
import api from '../api.js'
import { withRouter } from 'react-router'

export class EquipmentDetailsPage extends React.Component {
  state = {
    alert: null,
    ascending: false,
    data: {},
    error: null,
    loading: true,
    page: 0,
    perPage: 10,
    sortBy: 'updatedAt',
  }

  componentDidMount() {
    this.getData()
  }

  getData = () => {
    this.setState({ loading: true })
    const { ascending, page, perPage, sortBy } = this.state
    api.getShow(
      'equipment',
      this.props.match.params.id,
      {
        ascending,
        page,
        perPage,
        sortBy,
      },
      result => {
        this.setState({ loading: false, data: result })
      },
      error => {
        this.setState({ loading: false, error })
      },
    )
  }

  handleChange = object => {
    this.setState(object)
  }

  handleEditClick = () => {
    const id = this.state.data.equipment.id
    this.props.history.push(`/equipment/edit/${id}`)
  }

  handleDeleteClick = () => {
    if (confirm('Are you sure you want to delete this equipment?')) {
      const success = _response => {
        this.props.history.push('/')
      }

      const failure = error => {
        this.setState({ loading: false, alert: error })
      }

      api.deleteDestroy(
        'equipment',
        this.state.data.equipment.id,
        success,
        failure,
      )
    }
  }

  handleAddEventClick = () => {
    const id = this.state.data.equipment.id
    this.props.history.push(`/events/create/${id}`)
  }

  handlePageChange = (_event, newPage) => {
    this.setState({ page: newPage }, this.getData)
  }

  handlePerPageChange = event => {
    this.setState({ perPage: event.target.value }, this.getData)
  }

  handleSort = (sortBy, ascending) => {
    this.setState({ sortBy, ascending }, this.getData)
  }

  handleEventEditClick = id => {
    this.props.history.push(
      `/events/${this.state.data.equipment.id}/edit/${id}`,
    )
  }

  render() {
    if (this.state.loading) {
      return <Spinner />
    }

    if (this.state.error) {
      return <ErrorAlert closable={false} text={this.state.error} />
    }

    const equipment = this.state.data.equipment

    const headers = [
      { type: 'value', id: 'status', label: 'Status' },
      { type: 'value', id: 'jobNumber', label: 'Job Number' },
      { type: 'value', id: 'companyNotes', label: 'Company/Notes' },
      { type: 'value', id: 'startDate', label: 'Start Date' },
      { type: 'value', id: 'endDate', label: 'End Date' },
      { type: 'value', id: 'updatedAt', label: 'Updated At' },
      { type: 'button', id: 'edit' },
    ]

    const data = this.state.data.events.map(event => ({
      id: event.id,
      cells: [
        { id: 'status', type: 'value', value: event.status },
        { id: 'jobNumber', type: 'value', value: event.jobNumber },
        { id: 'companyNotes', type: 'value', value: event.companyNotes },
        { id: 'startDate', type: 'value', value: event.startDate },
        { id: 'endDate', type: 'value', value: event.endDate },
        { id: 'updatedAt', type: 'value', value: event.updatedAt },
        {
          id: 'edit',
          type: 'button',
          value: 'Edit',
          callback: this.handleEventEditClick,
        },
      ],
    }))

    return (
      <div>
        <Grid container>
          {this.state.alert && (
            <ErrorAlert closable={true} text={this.state.alert} />
          )}
          <Grid item xs={12}>
            <Title label={equipment.serialNumber} />
          </Grid>
          <Grid item xs={12}>
            OEM:
            <Button
              onClick={() => {
                this.props.history.push(`/oems/search/${equipment.oemName}`)
              }}
            >
              {equipment.oemName}
            </Button>
          </Grid>
          <Grid item xs={12}>
            Model:
            <Button
              onClick={() => {
                this.props.history.push(`/models/search/${equipment.modelName}`)
              }}
            >
              {equipment.modelName}
            </Button>
          </Grid>
          <Grid item xs={12}>
            Type:
            <Button
              onClick={() => {
                this.props.history.push(`/types/search/${equipment.typeName}`)
              }}
            >
              {equipment.typeName}
            </Button>
          </Grid>
          <Grid item xs={12}>
            Notes: {equipment.notes}
          </Grid>
          <Grid item xs={12}>
            Status: {equipment.eventStatus}
          </Grid>
          <Grid item xs={12}>
            Job Number: {equipment.eventJobNumber}
          </Grid>
          <Grid item xs={12}>
            Company/Notes: {equipment.eventCompanyNotes}
          </Grid>
          <Grid item xs={12}>
            Calibration Company: {equipment.calCompany}
          </Grid>
          <Grid item xs={12}>
            Calibration Due: {equipment.calDue}
          </Grid>
          <Grid item xs={12}>
            <Button onClick={this.handleEditClick}>Edit Equipment</Button>
          </Grid>
          <Grid item xs={12}>
            <Button onClick={this.handleDeleteClick}>Delete Equipment</Button>
          </Grid>
        </Grid>
        <Toolbar>
          <Typography variant='h5'>Events</Typography>
          <Button onClick={this.handleAddEventClick}>Add</Button>
        </Toolbar>
        <FullTable
          ascending={this.state.ascending}
          count={parseInt(this.state.data.count)}
          data={data}
          headers={headers}
          onPageChange={this.handlePageChange}
          onPerPageChange={this.handlePerPageChange}
          onSort={this.handleSort}
          page={this.state.page}
          perPage={this.state.perPage}
          perPageOptions={[5, 10, 25]}
          sortBy={this.state.sortBy}
        />
      </div>
    )
  }
}

export default withRouter(EquipmentDetailsPage)
