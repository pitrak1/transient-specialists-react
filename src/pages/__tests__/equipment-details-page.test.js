import React from 'react'
import { EquipmentDetailsPage } from '../equipment-details-page'
import { shallow } from 'enzyme'
import sinon from 'sinon'
import api from '../../api.js'

describe('EquipmentDetailsPage', () => {
  const data = {
    id: 1,
    serialNumber: 'some serail number',
    oemId: 2,
    oemName: 'OEM 2',
    modelId: 3,
    modelName: 'Model 3',
    typeId: 4,
    typeName: 'Type 4',
  }

  window.confirm = () => true

  it('renders spinner if not yet loaded', () => {
    const node = shallow(<EquipmentDetailsPage match={{ params: { id: 1 } }} />)
    expect(node).toMatchSnapshot()
  })

  it('renders an error if load fails', () => {
    api.getShow = sinon.stub().callsArgWith(3, 'some error')
    const node = shallow(<EquipmentDetailsPage match={{ params: { id: 1 } }} />)
    expect(node).toMatchSnapshot()
  })

  it('renders page if load succeeds', () => {
    api.getShow = sinon.stub().callsArgWith(2, data)
    const node = shallow(<EquipmentDetailsPage match={{ params: { id: 1 } }} />)
    expect(node).toMatchSnapshot()
  })

  it('redirects to equipment if submit succeeds', () => {
    api.getShow = sinon.stub().callsArgWith(2, data)
    api.deleteDestroy = sinon.stub().callsArg(2)
    const push = sinon.stub()
    const node = shallow(
      <EquipmentDetailsPage match={{ params: { id: 1 } }} history={{ push }} />,
    )
    node.instance().handleDeleteClick()
    expect(push.firstCall.args[0]).toBe('/')
  })

  it('displays alert if submit fails', () => {
    api.getShow = sinon.stub().callsArgWith(2, data)
    api.deleteDestroy = sinon.stub().callsArgWith(3, 'some error')
    const node = shallow(<EquipmentDetailsPage match={{ params: { id: 1 } }} />)
    node.instance().handleDeleteClick()
    expect(node).toMatchSnapshot()
  })
})
