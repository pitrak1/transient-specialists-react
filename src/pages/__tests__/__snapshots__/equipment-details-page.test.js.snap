// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EquipmentDetailsPage displays alert if submit fails 1`] = `
<div>
  <Alert
    liveRegionPoliteness="assertive"
    margin="x-small 0"
    open={true}
    screenReaderOnly={false}
    timeout={0}
    transition="fade"
    variant="error"
  >
    some error
  </Alert>
  <Heading
    border="none"
    color="inherit"
    ellipsis={false}
    level="h1"
    margin="medium"
  >
    some serail number
  </Heading>
  <div
    key="ID: "
  >
    <Text
      as="span"
      letterSpacing="normal"
      size="medium"
      weight="bold"
      wrap="normal"
    >
      ID: 
    </Text>
    <Text
      as="span"
      letterSpacing="normal"
      size="medium"
      wrap="normal"
    >
      1
    </Text>
  </div>
  <div
    key="Serial Number: "
  >
    <Text
      as="span"
      letterSpacing="normal"
      size="medium"
      weight="bold"
      wrap="normal"
    >
      Serial Number: 
    </Text>
    <Text
      as="span"
      letterSpacing="normal"
      size="medium"
      wrap="normal"
    >
      some serail number
    </Text>
  </div>
  <div
    key="OEM ID: "
  >
    <Text
      as="span"
      letterSpacing="normal"
      size="medium"
      weight="bold"
      wrap="normal"
    >
      OEM ID: 
    </Text>
    <Text
      as="span"
      letterSpacing="normal"
      size="medium"
      wrap="normal"
    >
      2
    </Text>
  </div>
  <div
    key="OEM Name: "
  >
    <Text
      as="span"
      letterSpacing="normal"
      size="medium"
      weight="bold"
      wrap="normal"
    >
      OEM Name: 
    </Text>
    <Text
      as="span"
      letterSpacing="normal"
      size="medium"
      wrap="normal"
    >
      OEM 2
    </Text>
  </div>
  <div
    key="Model ID: "
  >
    <Text
      as="span"
      letterSpacing="normal"
      size="medium"
      weight="bold"
      wrap="normal"
    >
      Model ID: 
    </Text>
    <Text
      as="span"
      letterSpacing="normal"
      size="medium"
      wrap="normal"
    >
      3
    </Text>
  </div>
  <div
    key="Model Name: "
  >
    <Text
      as="span"
      letterSpacing="normal"
      size="medium"
      weight="bold"
      wrap="normal"
    >
      Model Name: 
    </Text>
    <Text
      as="span"
      letterSpacing="normal"
      size="medium"
      wrap="normal"
    >
      Model 3
    </Text>
  </div>
  <div
    key="Type ID: "
  >
    <Text
      as="span"
      letterSpacing="normal"
      size="medium"
      weight="bold"
      wrap="normal"
    >
      Type ID: 
    </Text>
    <Text
      as="span"
      letterSpacing="normal"
      size="medium"
      wrap="normal"
    >
      4
    </Text>
  </div>
  <div
    key="Type Name: "
  >
    <Text
      as="span"
      letterSpacing="normal"
      size="medium"
      weight="bold"
      wrap="normal"
    >
      Type Name: 
    </Text>
    <Text
      as="span"
      letterSpacing="normal"
      size="medium"
      wrap="normal"
    >
      Type 4
    </Text>
  </div>
  <Button
    as="button"
    buttonRef={[Function]}
    cursor="pointer"
    fluidWidth={false}
    margin="0"
    onClick={[Function]}
    size="medium"
    type="button"
    variant="default"
  >
    Delete
  </Button>
</div>
`;

exports[`EquipmentDetailsPage renders an error if load fails 1`] = `
<Alert
  liveRegionPoliteness="assertive"
  margin="x-small 0"
  open={true}
  screenReaderOnly={false}
  timeout={0}
  transition="fade"
  variant="error"
>
  some error
</Alert>
`;

exports[`EquipmentDetailsPage renders page if load succeeds 1`] = `
<div>
  <Heading
    border="none"
    color="inherit"
    ellipsis={false}
    level="h1"
    margin="medium"
  >
    some serail number
  </Heading>
  <div
    key="ID: "
  >
    <Text
      as="span"
      letterSpacing="normal"
      size="medium"
      weight="bold"
      wrap="normal"
    >
      ID: 
    </Text>
    <Text
      as="span"
      letterSpacing="normal"
      size="medium"
      wrap="normal"
    >
      1
    </Text>
  </div>
  <div
    key="Serial Number: "
  >
    <Text
      as="span"
      letterSpacing="normal"
      size="medium"
      weight="bold"
      wrap="normal"
    >
      Serial Number: 
    </Text>
    <Text
      as="span"
      letterSpacing="normal"
      size="medium"
      wrap="normal"
    >
      some serail number
    </Text>
  </div>
  <div
    key="OEM ID: "
  >
    <Text
      as="span"
      letterSpacing="normal"
      size="medium"
      weight="bold"
      wrap="normal"
    >
      OEM ID: 
    </Text>
    <Text
      as="span"
      letterSpacing="normal"
      size="medium"
      wrap="normal"
    >
      2
    </Text>
  </div>
  <div
    key="OEM Name: "
  >
    <Text
      as="span"
      letterSpacing="normal"
      size="medium"
      weight="bold"
      wrap="normal"
    >
      OEM Name: 
    </Text>
    <Text
      as="span"
      letterSpacing="normal"
      size="medium"
      wrap="normal"
    >
      OEM 2
    </Text>
  </div>
  <div
    key="Model ID: "
  >
    <Text
      as="span"
      letterSpacing="normal"
      size="medium"
      weight="bold"
      wrap="normal"
    >
      Model ID: 
    </Text>
    <Text
      as="span"
      letterSpacing="normal"
      size="medium"
      wrap="normal"
    >
      3
    </Text>
  </div>
  <div
    key="Model Name: "
  >
    <Text
      as="span"
      letterSpacing="normal"
      size="medium"
      weight="bold"
      wrap="normal"
    >
      Model Name: 
    </Text>
    <Text
      as="span"
      letterSpacing="normal"
      size="medium"
      wrap="normal"
    >
      Model 3
    </Text>
  </div>
  <div
    key="Type ID: "
  >
    <Text
      as="span"
      letterSpacing="normal"
      size="medium"
      weight="bold"
      wrap="normal"
    >
      Type ID: 
    </Text>
    <Text
      as="span"
      letterSpacing="normal"
      size="medium"
      wrap="normal"
    >
      4
    </Text>
  </div>
  <div
    key="Type Name: "
  >
    <Text
      as="span"
      letterSpacing="normal"
      size="medium"
      weight="bold"
      wrap="normal"
    >
      Type Name: 
    </Text>
    <Text
      as="span"
      letterSpacing="normal"
      size="medium"
      wrap="normal"
    >
      Type 4
    </Text>
  </div>
  <Button
    as="button"
    buttonRef={[Function]}
    cursor="pointer"
    fluidWidth={false}
    margin="0"
    onClick={[Function]}
    size="medium"
    type="button"
    variant="default"
  >
    Delete
  </Button>
</div>
`;

exports[`EquipmentDetailsPage renders spinner if not yet loaded 1`] = `
<Spinner
  as="div"
  renderTitle="Loading"
  size="large"
  variant="default"
/>
`;
